File: zero-downtime-migration/src/routes/v2Users.ts
// Import Router from Express to create modular route handlers
import { Router } from "express";

// Import the User model to interact with the users table in the database
import User from "../models/User";

// Create a new Express router instance
const router = Router();

// ------------------------
// NEW API — Create a user
// ------------------------
router.post("/users", async (req, res) => {
  // Extract first name, last name, and email from the request body
  const { first_name, last_name, email } = req.body;

  // Combine first and last name into a single fullName string
  const fullName = `${first_name} ${last_name}`;

  // Create a new user in the database
  // Writes both individual names and fullName for convenience (dual write)
  const user = await User.create({
    firstName: first_name,  // Maps API first_name to DB column firstName
    lastName: last_name,    // Maps API last_name to DB column lastName
    fullName,               // Store full name in DB as well
    email                   // Store email
  });

  // Respond with the newly created user object as JSON
  res.json(user);
});

// ------------------------
// NEW API — Get a user by ID
// ------------------------
router.get("/users/:id", async (req, res) => {
  // Find user in the database by primary key (id from URL parameter)
  const user = await User.findByPk(req.params.id);

  // Respond with selected fields using optional chaining
  // Ensures no error if user is not found
  res.json({
    id: user?.id,              // User ID or undefined
    first_name: user?.firstName, // User firstName or undefined
    last_name: user?.lastName    // User lastName or undefined
  });
});

// Export the router so it can be mounted in the main Express app
export default router;

File: zero-downtime-migration/python/routes/v2_users.py
# routes/v2_users.py
# Defines version 2 of the Users API.
# This version uses the NEW schema (first_name / last_name),
# while still maintaining compatibility with the old schema
# through a dual-write strategy.

from fastapi import APIRouter
from database import SessionLocal
from models import User

# Create a router for v2 endpoints
router = APIRouter()

@router.post("/users")
def create_user_v2(first_name: str, last_name: str, email: str):
    # Create a new database session for this request
    db = SessionLocal()

    # Dual-write pattern:
    # New clients send first_name and last_name.
    # We also derive and write full_name so old systems,
    # reports, or background jobs continue to work.
    full_name = f"{first_name} {last_name}"

    # Create user with BOTH old and new columns populated
    user = User(
        first_name=first_name,
        last_name=last_name,
        full_name=full_name,  # legacy column
        email=email
    )

    # Persist the new user record
    db.add(user)
    db.commit()

    # Refresh instance to load generated fields (e.g., id)
    db.refresh(user)

    # Return response using ONLY the new schema
    return {
        "id": user.id,
        "first_name": user.first_name,
        "last_name": user.last_name
    }

@router.get("/users/{user_id}")
def get_user_v2(user_id: int):
    # Create a database session
    db = SessionLocal()

    # Fetch user by primary key
    user = db.query(User).filter(User.id == user_id).first()

    # Return response in the new API format
    return {
        "id": user.id,
        "first_name": user.first_name,
        "last_name": user.last_name
    }
